Sources/NIOTransportServices/FilterEmptyWritesHandler.swift

```diff
@@ -0,0 +1,145 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2020 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+// swift-tools-version:4.0
+//
+// swift-tools-version:4.0
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import NIO
+
+
+/// A `ChannelHandler` that checks for outbound writes of zero length, which are then dropped. This is
+/// due to a bug in `Network Framework`, where zero byte TCP writes lead to stalled connections.
+/// Write promises are confirmed in the correct order.
+public final class FilterEmptyWritesHandler: ChannelDuplexHandler {
+    public typealias InboundIn = ByteBuffer
+    public typealias InboundOut = ByteBuffer
+    public typealias OutboundIn = ByteBuffer
+    public typealias OutboundOut = ByteBuffer
+
+    private enum ChannelState {
+        case open
+        case closedFromRemote
+        case closedFromLocal
+        case error
+    }
+    
+    private var state: ChannelState = .closedFromLocal
+    private var prefixEmptyWritePromise: Optional<EventLoopPromise<Void>>
+    private var lastWritePromise: Optional<EventLoopPromise<Void>>
+    
+    public init() {
+        self.prefixEmptyWritePromise = nil
+        self.lastWritePromise = nil
+    }
+    
+    public func write(context: ChannelHandlerContext, data: NIOAny, promise: EventLoopPromise<Void>?) {
+        let buffer = self.unwrapOutboundIn(data)
```

Comment:
I think we need to check `self.state` here too. I think only if we're `.open` do we need to do what you're doing here, if we're not open we can just forward to the channel which should take care of failing them.

Reply:
That is very clever